lib/hypothesis/providers.rb

```diff
@@ -2,24 +2,90 @@
 
 module Hypothesis
   module Providers
+    def bits(n)
+      from_hypothesis_core HypothesisCoreBitProvider.new(n)
+    end
+
+    def composite(&block)
+      Hypothesis::Provider::Implementations::CompositeProvider.new(self, block)
+    end
+
     def integers
-      proc do |source|
-        if source.bits(1).positive?
-          source.bits(64)
+      composite do
+        if given(bits(1)).positive?
+          given(bits(64))
         else
           0
         end
       end
     end
 
     def strings
-      proc do |source|
-        if source.bits(1).positive?
+      composite do
+        if given(bits(1)).positive?
           'a'
         else
           'b'
         end
       end
     end
+
+    private
+
+    def from_hypothesis_core(core)
+      Hypothesis::Provider::Implementations::ProviderFromCore.new(
+        core
+      )
+    end
+  end
+
+  class Provider
+    module Implementations
+      class CompositeObject
+        def initialize(parent, source)
+          @parent = parent
+          @source = source
+        end
+
+        def given(provider)
+          @source.local_given(provider)
+        end
+
+        def assume(condition)
+          @source.assume(condition)
+        end
+
+        def respond_to_missing?(name)
+          @parent.respond_to? name
+        end
+
+        def method_missing(name, *args, &block)
+          @parent.send(name, *args, &block)
+        end
+      end
+
+      class CompositeProvider < Provider
+        def initialize(parent, block)
+          @parent = parent
+          @block = block
+        end
+
+        def provide(source)
+          CompositeObject.new(@parent, source).instance_eval(&@block)
```

Comment:
Instance eval is dark magic. Probably worth a comment explainig what you're doing with it.

Reply:
TBH the more I think about this API the more I hate it. I'm considering two alternatives:

* Adding `given` and `assume` methods to `Providers` so that `self` is the object including the modules. This requires some really awful state mangement but would allow for the same API with fewer weird pitfalls.
* Saying "fuck it, let's not go full Ruby" and having composite call `given` and `assume` on an argument passed to the block. I think this is the mature and responsible thing to do and I'm probably going to eventually talk myself into doing it.

Reply:
As you've probably seen, I've changed it so that `composite` just gets its source through its argument. I'm considering putting `given` and `assume` methods on the providers objects that could be used inside the `composite` body, but I'm resisting the idea for now. Worst case scenario it's easy to add those in later if I talk myself into it.